clients/src/main/java/org/apache/kafka/common/record/FileRecords.java

```diff
@@ -144,11 +154,12 @@ public FileRecords slice(int position, int size) throws IOException {
         if (size < 0)
             throw new IllegalArgumentException("Invalid size: " + size + " in read from " + this);
 
-        int end = this.start + position + size;
+        int sliceStart = this.start + position;
+        int sliceEnd = sliceStart + size;
         // handle integer overflow or if end is beyond the end of the file
-        if (end < 0 || end > start + currentSizeInBytes)
-            end = start + currentSizeInBytes;
-        return new FileRecords(file, channel, this.start + position, end, true);
+        if (sliceEnd < 0 || sliceEnd > start + currentSizeInBytes)
+            sliceEnd = start + currentSizeInBytes;
+        return new SliceRange(sliceStart, sliceEnd);
```

Comment:
I was thinking how we can simplify this a little bit and avoid the annoying object. I think the main logic here is validating the input and computing the size of the available bytes to read. 

```java
private int availableBytes(int position, int maxLength) {
   ...
   int limit = this.start + position + size;
   if (limit < 0 || limit > start + currentSizeInBytes)
     limit = start + currentSizeInBytes;
   return limit - position;
}
```
Then the caller can do something like this:
```
  int availableBytes = availableBytes(position, size);
  return new UnalignedFileRecords(channel, this.start + position, availableBytes);
```
What do you think? Any better?